File: src/main/kotlin/ps/mod/mapper/ModMapper.kt
package ps.mod.mapper

import leight.container.IContainer
import leight.mapper.AbstractMapper
import ps.mod.dto.ModDto
import ps.storage.module.mod.entity.ModEntity
import ps.storage.module.user.repository.lazyUserRepository
import ps.storage.module.vendor.repository.lazyVendorRepository
import ps.user.mapper.lazyUserMapper
import ps.vendor.mapper.lazyVendorMapper

class ModMapper(container: IContainer) : AbstractMapper<ModEntity, ModDto>(container) {
	private val userRepository by container.lazyUserRepository()
	private val userMapper by container.lazyUserMapper()
	private val vendorRepository by container.lazyVendorRepository()
	private val vendorMapper by container.lazyVendorMapper()

	override fun map(item: ModEntity) = ModDto.build {
		id = item.id
		name = item.name
		code = item.code
		power = item.power
		vendor = vendorMapper.map(vendorRepository.find(item.vendor))
		approvedBy = item.approvedBy?.let { userMapper.map(userRepository.find(it)) }
	}
}

fun IContainer.lazyModMapper() = lazy<ModMapper>()

File: src/PuffSmith/Mod/Mapper/ModMapper.php
<?php
declare(strict_types=1);

namespace PuffSmith\Mod\Mapper;

use Edde\Mapper\AbstractMapper;
use Edde\Tag\Mapper\TagMapperTrait;
use PuffSmith\Mod\Dto\ModDto;
use PuffSmith\Mod\Repository\ModTagRepositoryTrait;
use PuffSmith\Vendor\Mapper\VendorMapperTrait;
use PuffSmith\Vendor\Repository\VendorRepositoryTrait;
use function array_map;

class ModMapper extends AbstractMapper {
	use VendorRepositoryTrait;
	use VendorMapperTrait;
	use TagMapperTrait;
	use ModTagRepositoryTrait;

	public function item($item) {
		return $this->dtoService->fromArray(ModDto::class, [
			'id'          => $item->id,
			'name'        => $item->name,
			'power'       => $item->power,
			'voltage'     => $item->voltage,
			'vendorId'    => ($vendor = $this->vendorRepository->find($item->vendor_id))->id,
			'vendor'      => $this->vendorMapper->item($vendor),
			'cellTypes'   => $cellTypes = $this->tagMapper->map($this->modTagRepository->findByGroup($item->id, 'cell-type')),
			'cellTypeIds' => array_map(function ($cellType) {
				return $cellType->id;
			}, $cellTypes),
		]);
	}
}
